src/libraries/Microsoft.Extensions.DependencyInjection/src/ScopePool.cs

```diff
@@ -5,40 +5,42 @@
 using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Diagnostics;
-using System.Threading;
 using Microsoft.Extensions.DependencyInjection.ServiceLookup;
 
 namespace Microsoft.Extensions.DependencyInjection
 {
     internal class ScopePool
     {
-        // Modest number to re-use. We only really care about reuse for short lived scopes
-        private const int MaxQueueSize = 128;
-
-        private int _count;
-        private readonly ConcurrentQueue<State> _queue = new();
+        // Stash the scope state on a thread static
+        [ThreadStatic]
+        private static State t_tlsScopeState;
```

Comment:
What is the maximum size of this State object that will be stored here?

Reply:
I'd expect an extra large dictionary to be ~400 empty entries (that's what we saw in the orchard dumps).

Reply:
Is there an upper limit? The TLS caches like this typically have to have upper bond to prevent unbounded memory leak. Or they have to be kept in a weak handle so that they can get collected.

Reply:
I assume you're worried about the size of the entries empty entries on thread pool threads never being cleaned up?

Reply:
Right. 

For example, the dreadful StringBuilderCache has upper limit of 360 for the same reason: https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/Text/StringBuilderCache.cs#L12

Reply:
> Is there an upper limit in the worse case? 

I can add one, it's related to the comment here https://github.com/dotnet/runtime/blob/7fb611ec5aa9ef9c7080b20a3a386481f873bff6/src/libraries/Microsoft.Extensions.DependencyInjection/src/ScopePool.cs#L65. I'm not sure it matters though because in the most common cases, you end up constantly (per request) allocating a dictionary with that many entries anyways. It'll never get bigger than what your code was trying to use. I fear if we throw away entries that are too big then they'll be repeatedly allocated and thrown out. I could add a worst cast just in case (some extremely high number just to avoid pooling it at all).

> The TLS caches like this typically have to have upper bound to prevent unbounded memory leak. 

Aren't you intrinsically throttled by the number of threads? In practice, I don't see how this could lead to an unbounded memory leak.

Reply:
What about heteregonous requests? If you have one rarely executed request type that needs a lot of entries, all threads are eventually going to get the large collection attached to them and most of it will be useless.

>  I fear if we throw away entries that are too big then they'll be repeatedly allocated and thrown out. 

I do not see a problem with it. This optimization will not kick in and I think it is perfectly fine.

Reply:
> What about heterogonous requests? If you have one rarely executed request type that needs a lot of entries, all threads are eventually going to get the large collection attached to them and most of it will be useless.

Definitely possible, scopes can be used in the same app for background tasks and those might look different per request. I don't have any data but experience tells me the number of scoped services per request of the typical app is very similar. That's why I like the trimming capacity solution much better than throwing out the entire state object. It'll at the very least save initial resizes on other requests.

Reply:
I added a size. Not sure if it's too big and it's not a prime or power of 2 so it feels wrong 😄

Reply:
This makes it up to 24kB cached per thread if I have done the math right. Is it correct?

Reply:
How did you get 24KB? I've been trying to work out the size by looking at the size of an entry assuming it's also disposable:

![image](https://user-images.githubusercontent.com/95136/114520508-5e393680-9bf6-11eb-87d4-3cdc37811171.png)

Reply:
Actually, I think it is 36kB:
```
4 hashCode
4 int next

16 ServiceCacheKey (Type, Slot, padding)

8 object value

4 bucket
```

Reply:
Too big? 😄 I know people set large thread pool limits (1000s) but I think 1000 scoped services per request is pretty extreme, you probably have lots of memory anyways 😄

Reply:
Yes, it feels too big to me.

Reply:
OK I made some changes in the latest PR